### PR TITLE
Retry application-level error by default for datasets

### PR DESCRIPTION
**Pull request by jjyao**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -584,6 +584,11 @@ cdef execute_task(
                             core_worker.get_current_task_id())
                 except Exception as e:
                     is_application_level_error[0] = True
+                    if core_worker.get_current_task_retry_exceptions():
+                        logger.info("Application-level error was raised"
```
Comment (ericl) [inline python/ray/_raylet.pyx:588]:
```suggestion
                        logger.info("Task failed with retryable exception: {}"
```

File: python/ray/data/datasource/file_based_datasource.py [hunk 2]
```diff
@@ -119,7 +119,7 @@ def do_write(self,
 
         _write_block_to_file = self._write_block
 
-        @ray.remote
+        @ray.remote(retry_exceptions=True)
         def write_block(write_path: str, block: Block):
```
Comment (ericl) [inline python/ray/data/datasource/file_based_datasource.py:123]: Can you actually make this use cached_remote_fn, and move the retry exceptions flag into the cached_remote_fn fn?

File: python/ray/data/impl/compute.py [hunk 3]
```diff
@@ -37,7 +37,7 @@ def apply(self, fn: Any, remote_args: dict,
               blocks: BlockList[Any]) -> BlockList[Any]:
         map_bar = ProgressBar("Map Progress", total=len(blocks))
 
-        kwargs = remote_args.copy()
+        kwargs = {**{"retry_exceptions": True}, **remote_args}
```
Comment (ericl) [inline python/ray/data/impl/compute.py:40]: Let's move this into cached_remote_fn.

File: python/ray/data/read_api.py [hunk 4]
```diff
@@ -153,10 +153,14 @@ def read_datasource(datasource: Datasource[T],
...
+    default_ray_remote_args = {"retry_exceptions": True}
     if ray_remote_args:
-        remote_read = ray.remote(**ray_remote_args)(remote_read)
+        remote_read = ray.remote(**{
+            **default_ray_remote_args,
+            **ray_remote_args
+        })(remote_read)
     else:
-        remote_read = ray.remote(remote_read)
+        remote_read = ray.remote(**default_ray_remote_args)(remote_read)
```
Comment (ericl) [inline python/ray/data/read_api.py:163]: Move to cached_remote_fn

**Review by ericl — CHANGES_REQUESTED** [review]

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Oh one more thing, can you add a unit test?

**Comment by jjyao** [comment]
> Oh one more thing, can you add a unit test?

Yea, added one.

**Pull request merged**